src/Grids/Grids.jl

```diff
@@ -48,8 +48,19 @@ abstract type AbstractGrid{FT, TX, TY, TZ} end
 eltype(::AbstractGrid{FT}) where FT = FT
 topology(::AbstractGrid{FT, TX, TY, TZ}) where {FT, TX, TY, TZ} = (TX(), TY(), TZ())
 
-size(grid::AbstractGrid)   = (grid.Nx, grid.Ny, grid.Nz)
+x_topology(::AbstractGrid{FT, TX}) where {FT, TX} = TX()
+y_topology(::AbstractGrid{FT, TX, TY}) where {FT, TX, TY} = TY()
+z_topology(::AbstractGrid{FT, TX, TY, TZ}) where {FT, TX, TY, TZ} = TZ()
+
+size(grid::AbstractGrid) = (grid.Nx, grid.Ny, grid.Nz)
 length(grid::AbstractGrid) = (grid.Lx, grid.Ly, grid.Lz)
+halo_size(grid) = (grid.Hx, grid.Hy, grid.Hz)
+
+total_size(a) = size(a) # fallback
```

Comment:
Why do we need a fallback? Do we call `total_size` for anything but grids?

Reply:
We need a fallback so that it works equally well for raw `Array`s as well as fields.

Reply:
btw, this logic replicates widespread julia patterns; for example the function `parent` is defined on `Array`s (for which it is the identity function) in addition to wrappers (which return something different). We also define `data` in a similar way for fields, arrays, and array-like objects.

Reply:
That is true. I wasn't sure if it wasn't a quick utility for grids (in which case it might not make sense to generalize it to all types).